protocol/superc20/mirror-standard.md

```diff
@@ -0,0 +1,114 @@
+
+### Local functions
+
+Approvals functions will be implemented in a conventional way, specific to the contract, and not as mirror calls:
+
+- `approve(spender, amount)`
+- `allowance(owner, spender)`
+
+### Code sample
+
+Below, we show what the implementation of the mirror contract would look like.
+
+```solidity
+contract Mirror is Superc20 {
+  IERC20 public immutable LEGACY_TOKEN;
+  IL2StandardBridge public constant L2_STANDARD_BRIDGE = IL2StandardBridge(0x4200000000000000000000000000000000000010);
+
+  constructor(IERC20 _legacyToken) Superc20(_name, _symbol, _decimals) {
+    LEGACY_TOKEN = _legacyToken;
+    _name = sting.concat("Mirror",_legacyToken.name());
+    _symbol = string.concat("m:",_legacyToken.symbol());
+    _decimals = _legacyToken.decimals();
+  }
+
+  function totalSupply() public view override returns (uint256) {
+    return LEGACY_TOKEN.totalSupply();
+  }
+
+  function balanceOf(address account) public view override returns (uint256) {
+    return LEGACY_TOKEN.balanceOf(account);
+  }
+
+  function transfer(address recipient, uint256 amount) public override returns (bool) {
+    L2_STANDARD_BRIDGE.mirrorBurn(msg.sender, amount);
+    L2_STANDARD_BRIDGE.mirrorMint(recipient, amount);
+
+    emit Transfer(msg.sender, recipient, amount);
+    return true;
+  }
+
+  function transferFrom(address sender, address recipient, uint256 amount) public override returns (bool) {
+    uint256 currentAllowance = allowance(sender, msg.sender);
+    require(currentAllowance >= amount, "ERC20: transfer amount exceeds allowance");
+
+    _approve(sender, msg.sender, currentAllowance - amount);
+    L2_STANDARD_BRIDGE.mirrorBurn(sender, amount);
+    L2_STANDARD_BRIDGE.mirrorMint(recipient, amount);
+
+    emit Transfer(sender, recipient, amount);
+    return true;
+  }
+
+  // TODO: discuss possible events
+  function _mint(address _to, uint256 _amount) internal virtual override {
+    L2_STANDARD_BRIDGE.mirrorMint(_to, _amount);
+
+    emit Transfer(address(0), _to, _amount);
+    emit Mint(_to, _amount);
+  }
+
+  function _burn(address _from, uint256 _amount) internal virtual override {
+    L2_STANDARD_BRIDGE.mirrorBurn(_from, _amount);
+
+    emit Transfer(_from, address(0), _amount);
+    emit Burn(_from, _amount);
+  }
+}
+```
+
+### Deployment
+
+Deploying mirror versions will require introducing a new predeploy for`SuperchainERC20` factory. This factory would accomplish the following:
+
+- Designed to enable same-address deployments across OP Chains.
+- Address predictability for each token, we should know the corresponding Mirror.
+- Simple Factory code.
+- It may serve to create both `SuperchainERC20` standard tokens and mirror versions.
+
+### Upgrading `L2StandardBridge`
+
+We must update the `L2StandardBridge` to allow the mirror contract to mint and burn the mirrored legacy token. We can do this by adding a `mirrorMint()` and a `mirrorBurn()` functions that check the mirror is valid for the corresponding legacy token.
```

Comment:
How do we know this is secure? It just checks the L1 token matches? What is the exact check this does?

Reply:
The approach we were thinking to allow for proper validation here is to have the `L2StandardBridge` check if the address of the `msg.sender` (the mirror) corresponds to the correct mirror address deployed by the `Mirror` factory. In the `Mirror` factory design, each token is coupled with a single `Mirror`, so this can be done. I added a code example here: [Limited Mirror Factory](https://github.com/ethereum-optimism/design-docs/pull/37).

Here's an example code snippet:
```solidity
function mirrorMint(address _token, uint256 _amount) external {
	address _mirror = MIRROR_FACTORY.mirrors(_token);
	if (_mirror != msg.sender) revert();
	///...mint logic
}
```